invenio_records_resources/services/uow.py

```diff
@@ -0,0 +1,246 @@
+of Work maintains a list of operations and coordinates the commit, indexing and
+task execution.
+
+Note, this is NOT a clean implementation of the UoW design pattern. The
+main purpose of the Unit of Work in Invenio is to coordinate when the database
+transaction commit is called, and ensure tasks that has to run after the
+transcation is executed (such as indexing and running celery tasks).
+
+This ensures that we can group multiple services calls into a single database
+transaction and perform the necessary indexing/task execution afterwards.
+
+You should use the unit of work instead of running an explicit
+``db.session.commit()`` or ``db.session.rollback()`` in the code.
+You can still use all the remaining SQLAlchemy API such as ``db.session.add()``
+and friends.
+
+If you're only indexing (i.e. not database updates) also you don't need to use
+the unit of work.
+
+.. code-block:: python
+
+    with UnitOfWork(db.session):
+        # Be careful to always inject uow to the service. If not, the
+        # service will create its own UnitOfWork and commit.
+        service.communities.add(..., uow=uow)
+        service.publish(... , uow=uow)
+        uow.commit()
+
+
+When writing a service method that should commit a database transaction, you
+should use the provided decorator ``unit_of_work`` decorator (see example
+below):
+
+.. code-block:: python
+
+    @unit_of_work
+    def create(self, ... ,uow=None):
+        # ...
+        uow.register(RecordCommitOp(record, self.indexer))
+        # ...
+        # Do not use `db.session.commit()` in service.
+
+
+This ensures that the service method can be called with or without a unit of
+work. In case it's called without, the database transaction is automatically
+committed upon return from the function.
+"""
+
+
+from functools import wraps
+
+from invenio_db import db
+
+
+#
+# Unit of work operations
+#
+class Operation:
+    """Base class for unit of work operations."""
+
+    def on_register(self, uow):
+        """Called upon operation registration."""
+        pass
+
+    def on_commit(self, uow):
+        """Called in the commit phase (after the transaction is committed)."""
+        pass
+
+    def on_post_commit(self, uow):
+        """Called right after the commit phase."""
+        pass
+
+    def on_rollback(self, uow):
+        """Called in the rollback phase (after the transaction rollback)."""
+        pass
+
+    def on_post_rollback(self, uow):
+        """Called right after the rollback phase."""
+        pass
+
+
+class RecordCommitOp(Operation):
+    """Record commit operation with indexing."""
+
+    def __init__(self, record, indexer=None, index_refresh=False):
+        """Initialize the record commit operation."""
+        self._record = record
+        self._indexer = indexer
+        self._index_refresh = index_refresh
+
+    def on_register(self, uow):
+        """Commit record (will flush to the database)."""
+        self._record.commit()
```

Comment:
I am not sure I understand why we `commit` on `register`, but then `index` on `commit` below?